src/Illuminate/Http/Middleware/Http2ServerPush.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+namespace Illuminate\Http\Middleware;
+
+use Closure;
+use Illuminate\Http\Request;
+use Illuminate\Http\Response;
+use Symfony\Component\DomCrawler\Crawler;
+
+class Http2ServerPush
+{
+    /**
+     * The DomCrawler instance.
+     *
+     * @var \Symfony\Component\DomCrawler\Crawler
+     */
+    protected $crawler;
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param \Illuminate\Http\Request $request
+     * @param \Closure $next
+     *
+     * @return mixed
+     */
+    public function handle(Request $request, Closure $next, $limit = null)
+    {
+        $response = $next($request);
+
+        if ($response->isRedirection() || ! $response instanceof Response || $request->isJson()) {
+            return $response;
+        }
+
+        $this->generateAndAttachLinkHeaders($response, $limit);
+
+        return $response;
+    }
+
+    /**
+     * @param \Illuminate\Http\Response $response
+     *
+     * @return $this
+     */
+    protected function generateAndAttachLinkHeaders(Response $response, $limit = null)
+    {
+        $headers = $this->fetchLinkableNodes($response)
+            ->flatten(1)
+            ->map(function ($url) {
+                return $this->buildLinkHeaderString($url);
+            })
+            ->filter()
+            ->take($limit)
+            ->implode(',');
+
+        if (! empty(trim($headers))) {
+            $this->addLinkHeader($response, $headers);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Get the DomCrawler instance.
+     *
+     * @param \Illuminate\Http\Response $response
+     *
+     * @return \Symfony\Component\DomCrawler\Crawler
+     */
+    protected function getCrawler(Response $response)
+    {
+        if ($this->crawler) {
+            return $this->crawler;
+        }
+
+        return $this->crawler = new Crawler($response->getContent());
+    }
+
+    /**
+     * Get all nodes we are interested in pushing.
+     *
+     * @param \Illuminate\Http\Response $response
+     *
+     * @return \Illuminate\Support\Collection
+     */
+    protected function fetchLinkableNodes($response)
+    {
+        $crawler = $this->getCrawler($response);
+
+        return collect($crawler->filter('link, script[src], img[src]')->extract(['src', 'href']));
```

Comment:
The problem is that this picks up on everything, and only things from the same exact domain can be pushed. Not just anything in such tags.

Reply:
I think we could filter this list using the  `config('app.url')` to prevent pushing items from other domains. Thoughts?

Reply:
@JacobBennett That sounds wrong. The config entry is for url generation from artisan command. You could instead use the host from the $request instance.